src/plots/cartesian/layout_attributes.js

```diff
@@ -809,6 +809,21 @@ module.exports = {
             'vertically.'
         ].join(' ')
     },
+    autotickangles: {
+        valType: 'info_array',
+        freeLength: true,
+        items: {
+            valType: 'angle'
+        },
+        dflt: null,
```

Comment:
I am thinking here that the `dflt` value shouldn't be `null`.
IMHO it would be best to have `[0, 30, 90]` as the default.
@alexcjohnson do you agree?

Reply:
This is null to distinguish between user-set and unset autotickangles. If it's null, the old angle calculation will happen that used some heuristic. See line 3789 in `src/plots/cartesian/axes.js`

Reply:
I thought the old calculation was equivalent to `[30, 90]`, is it not? Actually though, it occurs to me perhaps we SHOULD include the `0` in this attribute, so that if you want to use the auto-rotate functionality but you want some rotation even when the labels don't overlap at all, you could do that. ie rather than always trying 0 first and only rotating if we need to, the logic would be that we always try the first angle in the list, then move to the next ones if we need to.

Reply:
Initially I set the default to [30, 90], but then I had failing tests: The old calculation sometimes chooses 90 degrees when the new calculation would choose 30.

Reply:
> Initially I set the default to [30, 90], but then I had failing tests: The old calculation sometimes chooses 90 degrees when the new calculation would choose 30.

I’m curious about this, because AFAICT your algorithm gets it right. What condition did this cause tests to fail? Maybe the test is wrong.

Also: I do think we should include the starting angle in the array, so that if for some reason you want to you can start with (for example) 30 degrees but still automatically increase to 90 degrees when necessary.

Otherwise this looks fantastic, we should be able to merge after addressing these last two points. 🌟